### ISSUE TITLE
`triu_/tril_` ignores memory overlaps which can lead to silent correctness issues

### ISSUE THREAD

**Issue by nikitaved** [pytorch/pytorch#165987]
### 🐛 Describe the bug

See the sample code below:

```python
In [2]: import torch                                                                                                                                                                        
                                                                                                                                                                                            
In [3]: x = torch.rand((), device="cuda")                                                                                                                                                   
                                                                                                                                                                                            
In [4]: x = x.expand(3, 3)                                                                                                                                                                  
                                                                                                                                                                                            
In [5]: x                                                                                                                                                                                   
Out[5]:                                                                                                                                                                                     
tensor([[0.3047, 0.3047, 0.3047],
        [0.3047, 0.3047, 0.3047],
        [0.3047, 0.3047, 0.3047]], device='cuda:0')

In [6]: x.triu_(1)
Out[6]: 
tensor([[0., 0., 0.],
        [0., 0., 0.],
        [0., 0., 0.]], device='cuda:0')
```
No warning, no error.

### Versions

current main

cc @malfet @jamesr66a

**Comment by agarg0627** [comment]
Hi, I opened #191854 for this, which builds on #165993 and #176782. It checks only the last two dimensions for overlap, so batch-dim overlap stays allowed, and it covers CPU, CUDA, and MPS.
